src/EditorFeatures/Core/Diagnostics/AbstractAggregateDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+/// <summary>
+/// Base type for all taggers that interact with the <see cref="IDiagnosticAnalyzerService"/> and produce tags for the
+/// diagnostics with different UI presentations.  It does no computation work itself, but instead defers that to it's
+/// underlying <see cref="SingleDiagnosticKindTaggerProvider"/>s.
+/// </summary>
+internal abstract partial class AbstractAggregateDiagnosticsTaggerProvider<TTag> : IViewTaggerProvider
+    where TTag : ITag
+{
+    /// <summary>
+    /// Underlying diagnostic tagger responsible for the syntax/semantic and compiler/analyzer split.
+    /// </summary>
+    private readonly ImmutableArray<SingleDiagnosticKindTaggerProvider> _diagnosticsTaggerProviders;
+
+    protected readonly IGlobalOptionService GlobalOptions;
+
+    protected AbstractAggregateDiagnosticsTaggerProvider(
+        IThreadingContext threadingContext,
+        IDiagnosticService diagnosticService,
+        IDiagnosticAnalyzerService analyzerService,
+        IGlobalOptionService globalOptions,
+        ITextBufferVisibilityTracker? visibilityTracker,
+        IAsynchronousOperationListener listener)
+    {
+        GlobalOptions = globalOptions;
+
+        _diagnosticsTaggerProviders = ImmutableArray.Create(
+            CreateDiagnosticsTaggerProvider(DiagnosticKind.CompilerSyntax),
+            CreateDiagnosticsTaggerProvider(DiagnosticKind.CompilerSemantic),
+            CreateDiagnosticsTaggerProvider(DiagnosticKind.AnalyzerSyntax),
```

Comment:
Just FYI that solution crawler based push diagnostics computed analyzer syntax diagnostics before compiler semantic ones. The order you have here also seems reasonable to me.

Reply:
Note, there is no ordering here. The individual sub taggers will tag concurrently.

Reply:
doc'ed this.